src/main/perl/Component.pm

```diff
@@ -278,6 +277,141 @@ sub _initialize
+(This does not affect the child module name.)
+
+=back
+
+Caveats
+
+=over
+
+=item NoActionSupported: this is declared by the (parent) component
+and applies for all redirections.
+
+=item C<LC::Exception> context (C<EC>): this is declared
+by the (parent) component, be careful when defining another one in the
+redirections.
+
+=back
+
+Example spma component has 'packager' configuration to switch between
+C<yum>, C<yumng>, C<ips> and C<apt>.
+
+The simplified implementation is as follows:
+
+The C<spma> component package contains
+    package NCM::Component::spma;
+    ...
+    use parent qw(NCM::Component);
+    use Readonly;
+    Readonly our $REDIRECT => {
+        name => 'packager',
+        lower => 1,
+        default => 'yum',
+    }
+
+This means that the module to use is based on the value
+of the C<packager> element in the component tree (based on
+C<prefix()>; in this case C</software/components/spma/packager>).
+In none is defined, the default C<yum> value is used.
+
+The actual module that is loaded from namespace based on the package name,
+with the packagename capitalised.
+In this example, by default this would be in C<NCM::Component::Spma>.
+However, with the C<lower> REDIRECT value true, the capitalisation is
+not performed and the resulting package is C<NCM::Component::spma::yum>.
+
+=cut
+
+sub _redirect
+{
+    my ($self, $method, $config) = @_;
+
+    my $whoami = ref($self);
+
+    my $redirect;
+
+    # TODO: prevent loop?
+    local $@;
+    eval {
+        my $varname = $whoami."::REDIRECT";
+        no strict 'refs';
+        $redirect = ${$varname};
+        use strict 'refs';
+    };
+    if ($@) {
+        $self->verbose("No REDIRECT defined for $whoami: $@");
+    }
+
+    if ($redirect) {
+        my $elname = $redirect->{name};
+        if (!defined($elname)) {
+            $self->error("Invalid REDIRECT for $whoami: missing name");
+            return;
+        }
+
+        # redirect current component
+        # based on spma.pm code call_entry_point
+        my $tree = $config->getTree($self->prefix());
+        my $childname = $tree->{$elname} || $redirect->{default};
+        if (!$childname) {
+            $self->error("No REDIRECT childname for $whoami found: ",
+                         "no name config (element $elname) and no REDIRECT default set");
+            return;
+        }
+
+        my @packagename = split(/::/, $whoami);
+        $packagename[-1] = ucfirst($packagename[-1]) if ! $redirect->{lower};
+        push(@packagename, $childname);
+
+        my $childpackagename = join('::', @packagename);
+
+        local $@;
+        my @warns;
+        eval {
+            local $SIG{__WARN__} = sub { push(@warns, $_[0]); };
+            load $childpackagename;
+        };
+        if ($@) {
+            $self->error("REDIRECT bad Perl code in $childpackagename: $@");
+            return;
+        }
+
+        # no real point in reporting the warnings on error
```

Comment:
I don't understand this comment. You say there's no real point doing this but do it anyway?

Reply:
i'll check again. i copied this code from somewhere else in ncm-ncd

Reply:
the remark is wrt reporting the warnings also in the previous error block. it won't add much and i can't just put this block before the error block, as it might mess up `$@`. so you only see the warnings when there's no error.